### PR TITLE
support for ordered choices, non-numbered choices by default, and basic support for \AMCnumericChoices

### PR DESCRIPTION
**Pull request by adl**
This implements supports for `\begin{choices}[o]` to disable shuffling, default to non-numbered answers as discussed in #18, and adds simple support for \AMCnumericChoices.

I'm a bit confused as to why `test.py` has the assertions about file equality commented. I suspect different LaTeX installations might produce different XML outputs?  Anyhow I've updated the xml files to use the output generated on my computer, after manually checking the differences.   Commit 00995f6 also includes changes to QCM.xml that come from some earlier commits.

Patch c2fb978 shows one issue in the conversion from LaTeX to XML, but that's unrelated to the patch itself.  It seems that when a question contains multiple paragraphs, they are just separated by a new line in the output instead of using `<p>...</p><p>...</p>`.  The question I've added to `test/QCM_wo-tikz.tex` is not rendered correctly in HTML because of that.

### CODE REVIEW

File: amc2moodle/amc2moodle/test/QCM_wo-tikz.xml [hunk 1]
```diff
@@ -181,5 +181,11 @@ A complicated chemical equation \(\ce{Hg^2+ -&gt;[I-] HgI2 -&gt;[I-] [Hg^{II}I4]
 
 
 
-]]></text></questiontext><answer format="html" fraction="100.0"><text><![CDATA[a simpler one \(\ce{CO2 + C -&gt; 2 CO}\).]]></text><fraction>100.0</fraction></answer><answer format="html" fraction="-50.0"><text><![CDATA[Wrong Choice! ]]></text><fraction>-50.0</fraction></answer><shuffleanswers>true</shuffleanswers><answernumbering>none</answernumbering><defaultgrade>1.0</defaultgrade></question>
+]]></text></questiontext><answer format="html" fraction="100.0"><text><![CDATA[a simpler one \(\ce{CO2 + C -&gt; 2 CO}\).]]></text><fraction>100.0</fraction></answer><answer format="html" fraction="-50.0"><text><![CDATA[Wrong Choice! ]]></text><fraction>-50.0</fraction></answer><defaultgrade>1.0</defaultgrade><shuffleanswers>true</shuffleanswers><answernumbering>none</answernumbering></question>
+  <question type="category"><category><text>$course$/test_notikz/bcl1</text></category></question>
+  <question type="numerical"><answer fraction="100"><text>290</text></answer><name><text>bcl1a</text></name><questiontext format="html"><text><![CDATA[
+Combien de fois le programme suivant affiche-t-il <tt>"x"</tt> ?
```
Comment (adl) [inline amc2moodle/amc2moodle/test/QCM_wo-tikz.xml:187]: I would really like to have a paragraph break here, like in the original LaTeX file.

Comment (adl) [reply amc2moodle/amc2moodle/test/QCM_wo-tikz.xml]: This is fixed by b2010b0.

**Comment by adl** [comment]
The fourth patch fixes the issue mentioned at the bottom of the PR.

**Comment by nennigb** [comment]
Hi, 

I will take time to integrate your modification next week. Just a quick comment to answer to your question :
> I'm a bit confused as to why test.py has the assertions about file equality

Depending on the version of image libraries, the compression not always return the same output. However, on the same computer the conversion is deterministic, thus, the match is not mandatory.  A message is printed when the match is ok. In the same way few other part of the file may change in the header depending latexml version.
There is several workaround, like removing few element in the match or separating question with image... but we didn't take time to work that.
These tests are useful to point out regression, but do not ensure that the conversion is the expected one. To do so, I use the moodle sandbox to upload and to test each new examples.

> `note` and `para`

Perhaps using `<ltx:text>` latexml environment is a better approach. I remember that I have already use it. I will investigate...


> `questionmultx` 
At the ends this kind of question could be convert as `<question type="numerical">`. I have planed to refactor the question processing asap.

Benoit

**Comment by adl** [comment]
Hi Benoit

> I will take time to integrate your modification next week. 

No problem.  I believe these were the main changes I need to start converting my question base to Moodle, but I plan to work on that as well next week.

> Depending on the version of image libraries, the compression not always return the same output. In the same way few other part of the file may change in the header depending latexml version.

Thanks for the explanations.

> These tests are useful to point out regression, but do not ensure that the conversion is the expected one. To do so, I use the moodle sandbox to upload and to test each new examples.

What I did was run `test.py`, then use `diff` to compare the generated files to the reference files and look for changes I expected to produce, then overwrite the reference files with the generated files.
 
I've also imported the XML files generated by the test suite into Moodle to make sure they look right.  (The only one that looked wrong is the mchem test, but that's to be expected on my school's installation.)

> > `note` and `para`
> 
> Perhaps using `<ltx:text>` latexml environment is a better approach. I remember that I have already use it. I will investigate...

I tried multiple things, including `<ltx:text>`, but I don't remember what was the issue with `<ltx:text>` in particular (I suspect it's just used for short strings and does not support paragraphs, but I might be wrong).  There seem to be some XML schema somewhere that disallows arbitrary nesting of XML tags, but when LaTeXML complains about such a situation, it just tells me that it's disallowed to use `<x>` in `<y>` without giving any hint about what I am allowed to put in `<y>` or which file specifies those constraints: this is really infuriating.

Since I could not locate the set of constraints, I've looked at how other commands were transformed for inspiration.  The `<ltx:note><ltx:inline-para>#1</ltx:inline-para></ltx:note>` pattern that I've ended-up using is actually how `\marginpar{}` is handled.

> > `questionmultx`
> 
> At the end, this kind of question could be convert into moodle `<question type="numerical">`. 

That could introduce errors for some people.  `\begin{questionmultx}` behaves exactly like `\begin{questionmult}` except it does not display the multiple-answers symbol.  It's possible to use `\begin{questionmultx}` with `\wrongchoice` and `\goodchoice` just to hide that symbol.  (I've done so in an exam where all questions were multiple-choice.)  Similarly, if one redefines `\multiSymbole` as empty then there is no difference between `\begin{questionmult}` and `\begin{questionmultx}` and one should be able to use `\AMCnumericalChoice{}` in `questionmult`.

(Now that I think of it, I don't know if it's possible to use `\AMCnumericalChoice` in a `\begin{question}` environment when only one digit needs to be selected.)

I believe the safest way to distinguish the "moodle question-type" is not only to look at the question environment, but also to look in the question at what kind of AMC answers are used.  That will be needed to implement `\AMCOpen` as well.

It would probably make more sense to decide the "moodle question-type" in the `xslt` transformation that occurs before the handling of grading and figures done by `convert.py` if this code is refactored using classes.

**Comment by adl** [comment]
I encountered one problem when converting:
```latex
\includegraphics{hilbert.png}
```
because it was output as
```xml
<file pathF="./hilbert.pn" ext="png" dim="width" size="200pt" name="hilbert" align="center"/>
```
Note the incorrect value for the `pathF` attribute.

Also in my case, this `\includegraphics` was in a numerical question, and in this case the picture was not encoded due to the check I added in c2fb978 to skip the grading code, since the conversion was mixed with the grading.

The latest patch (e5af8fb) addresses both issues.

**Comment by nennigb** [comment]
Ok, I can reproduce your problem with `\includegraphics`. Good catch !
Perhaps
```
img_path = os.path.dirname(os.path.join(pathin, img_name))
```
will be cleaner.

**Comment by adl** [comment]
> Perhaps
> 
> ```
> img_path = os.path.dirname(os.path.join(pathin, img_name))
> ```
> 
> will be cleaner.

That's essentially what e5af8fb does.  I've kept the `/./` -> `/` the replacement from the original code because I assumed it's useful in cases such as `img_name = "./figures/foo.png"` and `pathin = "test"`. This way we get `img_path = "test/figures"` instead of `img_path = "test/./figures"`.

**Comment by nennigb** [comment]
In fact `os.path` manage all these stuff natively :-). In my opinion, it is better to delegate all path manipulation to `os.path` for multiplatform support

**Comment by adl** [comment]
> In fact `os.path` manage all these stuff natively :-). In my opinion, it is better to delegate all path manipulation to `os.path` for multiplatform support

I don't get you.   The patch does this:

```python
>>> img_name = "./figures/foo.png"
>>> pathin = "test"
>>> os.path.dirname(os.path.join(pathin, img_name).replace('/./', '/'))
'test/figures'
```
If I understand correctly, you suggest:
```python
>>> os.path.dirname(os.path.join(pathin, img_name))
'test/./figures'
```
but the result is different.

**Comment by nennigb** [comment]
The string is different, but the path is the same
```
ls test/./Figures
or
ls test/Figures
```
If you want cleaner path, the best is to use
```
os.path.normpath('test/./Figures/4.png')
 'test/Figures/4.png'
os.path.normpath('test/./Figures/')
 'test/Figures'
os.path.normpath('test/./Figures')
 'test/Figures'
```
this yields
```
os.path.dirname(os.path.normpath(os.path.join(pathin, img_name)))
```

**Comment by adl** [comment]
> If you want cleaner path, 

Well, the `replace` was already there, so I'm assuming somebody else wanted that.  I'm just trying to not change behaviors unrelated to my issue.   But `normpath` is clearly better, I've updated the last patch to use it.

**Comment by nennigb** [comment]
I have use the replace when I was a newbie in python ;-).
I will merge your PR in this branch, make few modifications and merge it in the master.

**Comment by nennigb** [comment]
your PR is now merged in the master.

**Pull request merged**